cranelift/codegen/src/alias_analysis.rs

```diff
@@ -0,0 +1,396 @@
+//! runtime; this is a purely static concept. Instead, all
+//! memory-accessing instructions (loads and stores) are labeled with
+//! one of these four categories in the `MemFlags`. It is forbidden
+//! for a load or store to access memory under one category and a
+//! later load or store to access the same memory under a different
+//! category. This is ensured to be true by construction during
+//! frontend translation into CLIF and during legalization.
+//!
+//! Given that this non-aliasing property is ensured by the producer
+//! of CLIF, we can compute a *may-alias* property: one load or store
+//! may-alias another load or store if both access the same category
+//! of abstract state.
+//!
+//! The "last store" pass helps to compute this aliasing: it scans the
+//! code, finding at each program point the last instruction that
+//! *might have* written to a given part of abstract state.
+//!
+//! We can't say for sure that the "last store" *did* actually write
+//! that state, but we know for sure that no instruction *later* than
+//! it (up to the current instruction) did. However, we can get a
+//! must-alias property from this: if at a given load or store, we
+//! look backward to the "last store", *AND* we find that it has
+//! exactly the same address expression and type, then we know that
+//! the current instruction's access *must* be to the same memory
+//! location.
+//!
+//! To get this must-alias property, we compute a sparse table of
+//! "memory values": these are known equivalences between SSA `Value`s
+//! and particular locations in memory. The memory-values table is a
+//! mapping from (last store, address expression, type) to SSA
+//! value. At a store, we can insert into this table directly. At a
+//! load, we can also insert, if we don't already have a value (from
+//! the store that produced the load's value).
+//!
+//! Then we can do two optimizations at once given this table. If a
+//! load accesses a location identified by a (last store, address,
+//! type) key already in the table, we replace it with the SSA value
+//! for that memory location. This is usually known as "redundant load
+//! elimination" if the value came from an earlier load of the same
+//! location, or "store-to-load forwarding" if the value came from an
+//! earlier store to the same location.
+//!
+//! In theory we could also do *dead-store elimination*, where if a
+//! store overwrites a key in the table, *and* if no other load/store
+//! to the abstract state cateogry occurred, *and* no other trapping
+//! instruction occurred (at which point we need an up-to-date memory
+//! state because post-trap-termination memory state can be observed),
+//! *and* we can prove the original store could not have trapped, then
+//! we can eliminate the original store. Because this is so complex,
+//! and the conditions for doing it correctly when post-trap state
+//! must be correct likely reduce the potential benefit, we don't yet
+//! do this.
```

Comment:
File a follow up issue to track/discuss dead-store elimination?

Reply:
Sure, #4167.